File: library/src/main/java/org/fbme/lib/iec61499/declarations/FBInterfaceDeclaration.java
package org.fbme.lib.iec61499.declarations;

import org.fbme.lib.iec61499.descriptors.FBTypeDescriptor;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface FBInterfaceDeclaration extends NamedDeclaration {

    @NotNull List<EventDeclaration> getInputEvents();

    @NotNull List<EventDeclaration> getOutputEvents();

    @NotNull List<ParameterDeclaration> getInputParameters();

    @NotNull List<ParameterDeclaration> getOutputParameters();

    default @NotNull FBTypeDescriptor getTemplateTypeDescriptor() {
        if (this instanceof FBTypeDeclaration) {
            return ((FBTypeDeclaration) this).getTypeDescriptor();
        }
        if (this instanceof SubapplicationTypeDeclaration) {
            return ((SubapplicationTypeDeclaration) this).getTypeDescriptor();
        }
        if (this instanceof AdapterTypeDeclaration) {
            return ((AdapterTypeDeclaration) this).getSocketTypeDescriptor();
        }
        throw new IllegalArgumentException("Unknown declaration with FB interface: " + this.getClass().getName());
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/declarations/FBInterfaceDeclaration.kt
package org.fbme.lib.iec61499.declarations

import org.fbme.lib.common.Declaration
import org.fbme.lib.iec61499.descriptors.FBTypeDescriptor
import org.fbme.lib.iec61499.descriptors.SocketType

interface FBInterfaceDeclaration : Declaration {
    val inputEvents: MutableList<EventDeclaration>
    val outputEvents: MutableList<EventDeclaration>
    val inputParameters: MutableList<ParameterDeclaration>
    val outputParameters: MutableList<ParameterDeclaration>
    val templateTypeDescriptor: FBTypeDescriptor
        get() {
            if (this is FBTypeDeclaration) {
                return this.typeDescriptor
            }
            if (this is SubapplicationTypeDeclaration) {
                return this.typeDescriptor
            }
            if (this is AdapterTypeDeclaration) {
                return SocketType(this)
            }
            throw IllegalArgumentException("Unknown declaration with FB interface: " + this.javaClass.name)
        }

    fun isEmpty(): Boolean = inputEvents.isEmpty() &&
            outputEvents.isEmpty() &&
            inputParameters.isEmpty() &&
            outputParameters.isEmpty()

    fun copyEvents(declaration: FBInterfaceDeclaration) {
        inputEvents += declaration.inputEvents.map { it.copy() as EventDeclaration }
        outputEvents += declaration.outputEvents.map { it.copy() as EventDeclaration }
    }

    fun copyParameters(declaration: FBInterfaceDeclaration) {
        inputParameters += declaration.inputParameters.map { it.copy() as ParameterDeclaration }
        outputParameters += declaration.outputParameters.map { it.copy() as ParameterDeclaration }
    }
}
